feat: Add SQL DB assessment functionality and update SQL Server asses…

noyasarko opened the pull request:
The DB checks for now

closes #40

Commit:
feat: Add SQL DB assessment functionality and update SQL Server assessment

Nicwalle commented on SQL/SQLDBCheck.psm1:
```diff
@@ -0,0 +1,94 @@
+
+    # Checks if autoPauseDelay is enabled
+    [bool] hasAutopauseDelayEnabled() {
+        $DBSku = az sql db show --name $this.getDBName() --resource-group $this.getDBResourceGroup() --server $this.getDBServerName() --query 'sku' --output json | ConvertFrom-Json
```
You should not need to do an extra azure sql query. This is normally part of your `$this.DBObject` object. Hence, you would just need to get the `sku` from the DBObject

Nicwalle commented on SQL/SQLDBCheck.psm1:
```diff
@@ -0,0 +1,94 @@
+    [bool] hasAutopauseDelayEnabled() {
+        $DBSku = az sql db show --name $this.getDBName() --resource-group $this.getDBResourceGroup() --server $this.getDBServerName() --query 'sku' --output json | ConvertFrom-Json
+        if ($DBSku.tier -eq 'GeneralPurpose' -and $DBSku.family -eq 'Gen5') {
+            $DBProperties = az sql db show --name $this.getDBName() --resource-group $this.getDBResourceGroup() --server $this.getDBServerName() --query 'autoPauseDelay' --output json | ConvertFrom-Json
```
Same here, this information should be part of the `$this.DBObject` object

Nicwalle commented on SQL/SQLDBCheck.psm1:
```diff
@@ -0,0 +1,94 @@
+
+    # Checks if Back Ups are enabled
+    [bool] hasBackUpsEnabled() {
+        $DBbackupJson = az sql db ltr-backup list --location $this.getDBLocation() --resource-group $this.getDBResourceGroup() --server $this.getDBServerName() | ConvertFrom-Json
```
To be discussed with Frederik but I'm wondering if it is the right approach as backups might be stored in a location that is different from the one of the database (will then fail the test while backups can exist). Lacking knowledge and maybe it will not be the case but to be validated by Frederik

Nicwalle replied:
@noyasarko have you been able to discuss this one with Frederik?

Nicwalle replied:
Will be discussed later on. Resolving for now

Nicwalle commented on SQL/SQLDBCheck.psm1:
```diff
@@ -0,0 +1,94 @@
+
+    #Checks if there are replica's configured
+    [bool] hasReplicasConfigured(){
+        $DBReplica = az sql db replica list-links --name $this.getDBName --server $this.getDBServerName --resource-group $this.getDBResourceGroup 
```
I've tested it with a database with no replicas and it returns and empty list `[]` not null. 

Also, you miss the parentheses `()` after the getters: getDBName() etc.

Nicwalle commented on SQL/SQLDBCheck.psm1:
```diff
@@ -0,0 +1,94 @@
+    }
+
+    #Checks database size REWORK THIS CHECK 
+    [bool] isDatabaseSizeWithinLimit(){
+        $databaseSizeInMB = az sql db show --name $this.getDBName --server $this.getDBServerName --resource-group $this.getDBResourceGroup --query maxSizeBytes --output tsv | ForEach-Object { [int]$_ / 1MB }
+        $limitInMB = 5000 #check currently set max set limit, more than 90% the check should fail
+        return $databaseSizeInMB -le $limitInMB
+    }
```
To keep the code clean, I would comment (or completely remove) this check as long as it does not work.

Nicwalle commented on SQL/SQLDBCheck.psm1:
```diff
@@ -0,0 +1,94 @@
+    [CheckResults] assess() {
+        $rules = Get-Content SQL/sqlDBRules.json | ConvertFrom-Json
+
+        $this.Results.Add("Name", $this.getDBName())
+        $this.Results.Add("Resource_Group", $this.getDBResourceGroup())
```
For better analysis of the results, could you also add the server name, location and any other information that you think is relevant for the output csv file

Nicwalle commented on SQL/SQLServerCheck.psm1:
```diff
@@ -83,7 +83,7 @@ class SQLServerCheck: ResourceCheck {
+        $rules = Get-Content SQL/sqlServerRules.json | ConvertFrom-Json
 
         $this.Results.Add("Name", $this.getServerName())
         $this.Results.Add("Resource_Group", $this.getServerResourceGroup())
```
Can you add the location to the results so it shows up in the csv output

Nicwalle commented on SQL/StartSQLAssessment.ps1:
```diff
@@ -18,15 +18,18 @@ foreach ($currentSubscription in $subscriptions) {
+        Write-Host ""
+
+        #$jsonSQLDBs = az sql db list --resource-group NoyaSQLdb --server noyaserver -o json --only-show-errors | ConvertFrom-Json | Where-Object { $_.name -ne 'master' } | ConvertTo-Json
+        $jsonSQLDBs = az sql db list --resource-group $currentserver.resourceGroup --server $currentserver.name -o json --only-show-errors | ConvertFrom-Json | Where-Object { $_.name -ne 'master' } | ConvertTo-Json
```
Can you add the logic to same the output of this command to a file `sql_db_raw_$today.json`

Commit:
feat: Refactor SQLDBCheck and SQLServerCheck classes for improved readability and maintainability

Nicwalle commented:
Hi @noyasarko did you already make some progress on this one. If so, mind pushing it so we can merge the branch. Thanks

Commit:
feat: Improve database size check in SQLDBCheck class

The commit refactors the `isDatabaseSizeWithinLimit` method in the `SQLDBCheck` class to improve readability and maintainability. It now checks if the database size exceeds 90% of the limit by retrieving the current size and maximum size of the database using the `az sql db show` command. The method also handles cases where the size information is not available.

Nicwalle commented on SQL/SQLDBCheck.psm1:
```diff
@@ -0,0 +1,105 @@
+
+    # Checks if autoPauseDelay is enabled
+    [bool] hasAutopauseDelayEnabled() {
+        if ($this.sku.tier -eq 'GeneralPurpose' -and $this.sku.family -eq 'Gen5') {
```
$this.sku.tier does not exist. sku.tier is part of the "$this.DBObject"

Nicwalle commented on SQL/SQLDBCheck.psm1:
```diff
@@ -0,0 +1,105 @@
+    [string] getDBServerName() {
+        #DBObject needs to access its server's name
+        return $this.getServerName()
+    }
```
This function is not required, it just returns the content of the getServerName function.

Nicwalle commented on SQL/SQLDBCheck.psm1:
```diff
@@ -0,0 +1,105 @@
+        $rules = Get-Content SQL/sqlDBRules.json | ConvertFrom-Json
+
+        $this.Results.Add("Name", $this.getDBName())
+        $this.Results.Add("Server", $this.getDBServerName())
```
Please, use the already existing getServerName

Nicwalle commented on SQL/SQLServerCheck.psm1:
```diff
@@ -83,10 +87,11 @@ class SQLServerCheck: ResourceCheck {
 
         $this.Results.Add("Name", $this.getServerName())
         $this.Results.Add("Resource_Group", $this.getServerResourceGroup())
+        $this.Results.Add("Location", $this.getServerLocation)
```
You miss the parentheses

Nicwalle commented on SQL/SQLDBCheck.psm1:
```diff
@@ -0,0 +1,105 @@
+
+    # Checks if Auditing on server level is enabled
+    [bool] hasAuditingDBEnabled() {
+        $DBAuditing =  az sql db audit-policy show --name $this.getDBName --resource-group $this.getDBResourceGroup() --server $this.getServerName() --query 'state' -o json | ConvertFrom-Json 
```
Missing parentheses with the getDBName().
FYI, this shows up in the transcript:
````
ERROR: (ParentResourceNotFound) Failed to perform 'read' on resource(s) of type 'servers/databases/auditingSettings', because the parent resource '/subscriptions/[REDACTED]/resourceGroups/[REDACTED]/providers/Microsoft.Sql/servers/[REDACTED]/databases/string getDBName()' could not be found.
Code: ParentResourceNotFound
Message: Failed to perform 'read' on resource(s) of type 'servers/databases/auditingSettings', because the parent resource '/subscriptions/[REDACTED]/resourceGroups/[REDACTED]/providers/Microsoft.Sql/servers/[REDACTED]/databases/string getDBName()' could not be found.
```

Nicwalle committed:
Changed JSON output from PSObject to JSON

Commit:
Refactor SQLDBCheck and SQLServerCheck classes for improved readability and maintainability

Commit:
Merge branch '40-sql-db-check' of https://github.com/beluxappdev/health-check into 40-sql-db-check

Nicwalle commented on SQL/SQLDBCheck.psm1:
```diff
@@ -0,0 +1,100 @@
+
+    #Checks if there are replica's configured
+    [bool] hasReplicasConfigured(){
+        $DBReplica = az sql db replica list-links --name $this.getDBName() --server $this.getServerName() --resource-group $this.getDBResourceGroup() 
+        return $DBReplica -gt 0
```
$DBReplica contains a list. Empty (`[]`) if no replicas are defined, and non-empty otherwise. You cannot compare a list with a number. You need to check the length of the list.

Nicwalle commented on SQL/SQLDBCheck.psm1:
```diff
@@ -0,0 +1,100 @@
+
+
+    # Checks if the database size exceeds 90% of the limit
+    [bool] isDatabaseSizeWithinLimit() {
+        $databaseSizeInBytes = az sql db show --name $this.getDBName() --server $this.getServerName() --resource-group $this.getDBResourceGroup() --query 'properties.currentServiceObjectiveName' --output tsv
+        $maxSizeInBytes = az sql db show --name $this.getDBName() --server $this.getServerName() --resource-group $this.getDBResourceGroup() --query 'properties.maxSizeBytes' --output tsv
+
+        if ($databaseSizeInBytes -eq $null -or $maxSizeInBytes -eq $null) {
+            return $false
+        }
+
+        $databaseSizeInMB = [math]::Round($databaseSizeInBytes / 1MB, 2)
+        $maxSizeInMB = [math]::Round($maxSizeInBytes / 1MB, 2)
+        $threshold = $maxSizeInMB * 0.9
+
+        return $databaseSizeInMB -le $threshold
+    }
```
This check does not seem to be called. Is it ready to be used? If so, can you please optimize by leveraging the values of the $this.DBObject instead of doing extra calls with the `az sql db show` command

noyasarko replied:
yes indeed, this is a WIP and will be discussed with Frederik, I have it here just in case

Nicwalle replied:
Will be discussed later on. Resolving conversation for now

Commit:
Refactor SQLDBCheck hasReplicasConfigured method for improved accuracy

The commit modifies the `hasReplicasConfigured` method in the `SQLDBCheck` class to improve the accuracy of checking if there are replicas configured for the database. Instead of checking if the result of `az sql db replica list-links` is greater than 0, it now checks if the count of the result is greater than 0. This ensures that the method correctly identifies if there are replicas configured.

Nicwalle committed:
Fixed replica check

Commit:
Refactor SQLDBCheck and SQLServerCheck classes for improved readability and maintainability

Nicwalle committed:
renamed readme file
